File: backend/src/controllers/auth.controller.js
const jwt = require('jsonwebtoken');
const { validationResult } = require('express-validator');
const User = require('../models/User');
const HealthRecord = require('../models/HealthRecord');

const signToken = (userId) =>
  jwt.sign({ id: userId }, process.env.JWT_SECRET || 'myheath_jwt_secret', {
    expiresIn: '7d',
  });

exports.register = async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { firstName, lastName, email, password, role, phone, specialty, gender, hasDiabetes, diabetesType } =
      req.body;

    const exists = await User.findOne({ email: email.toLowerCase() });
    if (exists) {
      return res.status(409).json({ message: 'This email is already registered' });
    }

    const allowedRoles = ['patient', 'doctor'];
    const safeRole = allowedRoles.includes(role) ? role : 'patient';
    const safeGender = gender === 'man' ? 'man' : 'woman';

    const user = await User.create({
      firstName,
      lastName,
      email,
      password,
      role: safeRole,
      gender: safeGender,
      hasDiabetes: Boolean(hasDiabetes),
      diabetesType: hasDiabetes ? diabetesType || 'type2' : 'none',
      phone: phone || '',
      specialty:
        safeRole === 'doctor'
          ? specialty || (safeGender === 'man' ? 'General medicine' : 'Gynecology')
          : '',
    });

    if (user.role === 'patient') {
      await HealthRecord.create({ patient: user._id });
    }

    const token = signToken(user._id);
    res.status(201).json({ token, user: user.toSafeJSON() });
  } catch (error) {
    console.error('register error:', error);
    res.status(500).json({ message: 'Registration failed' });
  }
};

exports.login = async (req, res) => {
  try {
    const { email, password } = req.body;
    if (!email || !password) {
      return res.status(400).json({ message: 'Email and password are required' });
    }

    const user = await User.findOne({ email: email.toLowerCase() }).select('+password');
    if (!user || !(await user.comparePassword(password))) {
      return res.status(401).json({ message: 'Invalid credentials' });
    }

    const token = signToken(user._id);
    res.json({ token, user: user.toSafeJSON() });
  } catch (error) {
    console.error('login error:', error);
    res.status(500).json({ message: 'Login failed' });
  }
};

exports.me = async (req, res) => {
  res.json({ user: req.user.toSafeJSON() });
};

exports.listDoctors = async (_req, res) => {
  try {
    const doctors = await User.find({ role: 'doctor', isActive: true }).select(
      'firstName lastName specialty email'
    );
    res.json({ doctors });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load doctors' });
  }
};

exports.assignDoctor = async (req, res) => {
  try {
    const { doctorId } = req.body;
    const doctor = await User.findOne({ _id: doctorId, role: 'doctor' });
    if (!doctor) {
      return res.status(404).json({ message: 'Doctor not found' });
    }

    req.user.assignedDoctor = doctor._id;
    await req.user.save();
    res.json({ user: req.user.toSafeJSON() });
  } catch (error) {
    res.status(500).json({ message: 'Failed to assign doctor' });
  }
};

exports.listUsers = async (_req, res) => {
  try {
    const users = await User.find({ isActive: true })
      .select('firstName lastName email role gender specialty hasDiabetes createdAt')
      .sort({ role: 1, lastName: 1 });

    const counts = {
      total: users.length,
      admin: users.filter((u) => u.role === 'admin').length,
      doctor: users.filter((u) => u.role === 'doctor').length,
      patient: users.filter((u) => u.role === 'patient').length,
    };

    res.json({ counts, users });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load users' });
  }
};

File: backend-dotnet/MyHeath.Api/Controllers/AuthController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using MyHeath.Api.Infrastructure;
using MyHeath.Api.Models;
using MyHeath.Api.Services;

namespace MyHeath.Api.Controllers;

[ApiController]
[Route("api/auth")]
public class AuthController : ApiControllerBase
{
    private readonly JwtTokenService _jwt;

    public AuthController(MongoContext db, JwtTokenService jwt) : base(db) => _jwt = jwt;

    public record RegisterDto(
        string FirstName, string LastName, string Email, string Password,
        string? Role, string? Phone, string? Specialty, string? Gender,
        bool? HasDiabetes, string? DiabetesType);

    public record LoginDto(string Email, string Password);
    public record AssignDoctorDto(string DoctorId);

    [HttpPost("register")]
    [AllowAnonymous]
    public async Task<IActionResult> Register([FromBody] RegisterDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.Email) || string.IsNullOrWhiteSpace(dto.Password) || dto.Password.Length < 6)
            return Fail("Invalid registration data");

        var email = dto.Email.Trim().ToLowerInvariant();
        if (await Db.Users.Find(u => u.Email == email).AnyAsync())
            return Fail("This email is already registered", 409);

        var role = dto.Role is "doctor" or "patient" ? dto.Role! : Roles.Patient;
        var gender = dto.Gender == "man" ? "man" : "woman";
        var user = new User
        {
            FirstName = dto.FirstName.Trim(),
            LastName = dto.LastName.Trim(),
            Email = email,
            Password = BCrypt.Net.BCrypt.HashPassword(dto.Password, 12),
            Role = role,
            Gender = gender,
            HasDiabetes = dto.HasDiabetes == true,
            DiabetesType = dto.HasDiabetes == true ? (dto.DiabetesType ?? "type2") : "none",
            Phone = dto.Phone ?? "",
            Specialty = role == Roles.Doctor
                ? (dto.Specialty ?? (gender == "man" ? "General medicine" : "Gynecology"))
                : ""
        };
        await Db.Users.InsertOneAsync(user);

        if (user.Role == Roles.Patient)
        {
            await Db.HealthRecords.InsertOneAsync(new HealthRecord { Patient = user.Id });
        }

        return StatusCode(201, new { token = _jwt.CreateToken(user), user = UserMapper.ToSafe(user) });
    }

    [HttpPost("login")]
    [AllowAnonymous]
    public async Task<IActionResult> Login([FromBody] LoginDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.Email) || string.IsNullOrWhiteSpace(dto.Password))
            return Fail("Email and password are required");

        var email = dto.Email.Trim().ToLowerInvariant();
        var user = await Db.Users.Find(u => u.Email == email).FirstOrDefaultAsync();
        if (user is null || !user.IsActive || !BCrypt.Net.BCrypt.Verify(dto.Password, user.Password))
            return Fail("Invalid credentials", 401);

        return Ok(new { token = _jwt.CreateToken(user), user = UserMapper.ToSafe(user) });
    }

    [HttpGet("me")]
    [Authorize]
    public async Task<IActionResult> Me()
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        return Ok(new { user = UserMapper.ToSafe(user) });
    }

    [HttpGet("doctors")]
    [Authorize]
    public async Task<IActionResult> ListDoctors()
    {
        var doctors = await Db.Users.Find(u => u.Role == Roles.Doctor && u.IsActive)
            .Project(u => new { u.FirstName, u.LastName, u.Specialty, u.Email, id = u.Id })
            .ToListAsync();
        // Match Node shape (firstName etc.) — project loses Id naming; rebuild
        var list = await Db.Users.Find(u => u.Role == Roles.Doctor && u.IsActive).ToListAsync();
        return Ok(new
        {
            doctors = list.Select(d => new
            {
                _id = d.Id,
                id = d.Id,
                firstName = d.FirstName,
                lastName = d.LastName,
                specialty = d.Specialty,
                email = d.Email
            })
        });
    }

    [HttpPost("assign-doctor")]
    [Authorize(Roles = Roles.Patient)]
    public async Task<IActionResult> AssignDoctor([FromBody] AssignDoctorDto dto)
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        var doctor = await Db.Users.Find(u => u.Id == dto.DoctorId && u.Role == Roles.Doctor).FirstOrDefaultAsync();
        if (doctor is null) return Fail("Doctor not found", 404);

        user.AssignedDoctor = doctor.Id;
        user.UpdatedAt = DateTime.UtcNow;
        await Db.Users.ReplaceOneAsync(u => u.Id == user.Id, user);
        return Ok(new { user = UserMapper.ToSafe(user) });
    }
}
